dokka-runners/dokka-gradle-plugin/src/main/kotlin/internal/PluginFeaturesService.kt

```diff
@@ -0,0 +1,144 @@
+        val suppressV2PluginMessage: Property<Boolean>
+    }
+
+    internal val enableV2Plugin: Boolean get() = parameters.enableV2Plugin.get()
+    private val suppressV2PluginMessage: Boolean get() = parameters.suppressV2PluginMessage.get()
+
+    /** Used to only log the V1 message once per project, regardless of how many subprojects there are. */
+    private var v1WarningLogged: Boolean = false
+
+    /** Used to only log the V2 message once per project, regardless of how many subprojects there are. */
+    private var v2MessageLogged: Boolean = false
+
+    fun logV1Warning() {
+        if (!v1WarningLogged) {
+            logger.warn(
+                """
+                |⚠ Warning: Dokka Gradle Plugin V1 mode is enabled
+                |
+                |  V1 mode is deprecated, and will be removed in Dokka version 2.0.20.
+                |
+                |  Please migrate Dokka Gradle Plugin to V2. This will require updating your project.
+                |  To get started check out the Dokka Gradle Plugin Migration guide
+                |      https://kotl.in/dokka-gradle-migration
+                |
+                |  Once you have prepared your project, enable V2 by adding
+                |      $V2_PLUGIN_ENABLED_FLAG=true
+                |  to your project's `gradle.properties`
+                |
+                |  Please report any feedback or problems to Dokka GitHub Issues
+                |      https://github.com/Kotlin/dokka/issues/
+                """.trimMargin().surroundWithBorder()
+            )
+            v1WarningLogged = true
+        }
+    }
+
+    fun logV2Message() {
+        if (!suppressV2PluginMessage && !v2MessageLogged) {
+            logger.lifecycle(
+                """
+                |Dokka Gradle Plugin V2 is enabled ♡
+                |
+                |  We would appreciate your feedback!
+                |  Please report any feedback or problems to Dokka GitHub Issues
+                |      https://github.com/Kotlin/dokka/issues/
+                |
+                |  If you need help or advice, check out the migration guide
+                |      https://kotl.in/dokka-gradle-migration
+                |
+                |  You can suppress this message by adding
+                |      $V2_PLUGIN_ENABLED_QUIET_FLAG=true
+                |  to your project's `gradle.properties`
+                """.trimMargin().surroundWithBorder()
+            )
+            v2MessageLogged = true
+        }
+    }
+
+    companion object {
+        private val logger = Logging.getLogger(PluginFeaturesService::class.java)
+
+        /**
+         * Register a new [PluginFeaturesService], or get an existing instance.
+         */
+        internal val Project.pluginFeaturesService: PluginFeaturesService
+            get() = gradle.sharedServices
+                .registerIfAbsent("PluginFeaturesService.internal", PluginFeaturesService::class) {
```

Comment:
TIL, thanks.

Reply:
I'm not sure what the purpose of adding the classloader hashcode is though? What problem does it prevent?

Reply:
I had a good conversation with `@ALikhachev` about this, and I understand better now. 

I updated MigrationMessagesTest to recreate the issue (removed DGP from the root project, and make the plugins in each subproject different). I then implemented the workaround from KGP.

To ensure the messages are still only logged once I updated `PluginFeaturesService` registration so that if [the classloader bug](https://github.com/gradle/gradle/issues/17559) happens then the KGP workaround will kick-in, but the registered service won't log any messages.